### ISSUE TITLE
Project Status

### ISSUE THREAD

**Issue by bradstewart** [flegall/monopack#35]
Glad to see someone working on this! I was about to start writing a similar build system when I stumbled across this project. 

From glancing at the source, it seems you've made some great progress. Do you have a second to summarize what's been implemented/what's left to do?

If you're interested in some help, I'm happy to pitch in.

**Comment by flegall** [comment]
Hi !

Thanks for the feedback and interest on the project !

Here are the releases I've planned : 

v0.1:
- includes the **build** command.

v1.0:
- includes the **run** and **debug** command
- includes **--watch** toggle for build/run/debug commands using nodemon's api.



We're not so far from a v0.1, I'd say it'll ready within next week !



Here is a todo list for v0.1 (make the **build** command run)
- [X] Implement basic CLI options with yargs.
- [X] Implement builder with webpack
- [ ] Implement dependency collection algorithm: 90% is done
    -  The core dependency collection algorithm is imple
mented
    - There are a few bugs here and here (monopack cannot build itself for now due to the fact it doesn't handle package json with no dependency fields). I have to write a few tests to fill those bugs.
- [ ] Handle all the dependency collection result cases to display appropriate message (https://github.com/flegall/monopack/blob/master/packages/monopack-cli/src/main.js#L97)
- [ ] Copy collected yarn.lock file
- [X] Invoke yarn after building

Here is a todo list for v1.0 : 
- [ ] Implement run command
- [ ] Implement debug command
- [ ] Implement --watch mode for all commands


For later on, I'm thinking of the following features : 
- Dockerfile generation
- Build with rollup:
    - actually, Rollup's API makes it possible to run the dependency collection algorithm, as it also exposes an externals decide() function)
    - Rollup would make tree shaking possible without relying on uglifyJS. Of course for node.js bundle's tree shaking is not as critical as in web bundles, but nevertheless it's a nice-to-have.
- Add an option to skip yarn installation (useless if you build in a build/target dir within your monorepo).
- Include a typescript loader in webpack default config.

**Comment by flegall** [comment]
For help, I would love to have :

- design/code reviews (if you have time for it, I'd love that to have your point of view on the code, especially the parts that are too complicated to understand so I could focus effort on rewriting/simplifying those parts).
- code contribution on v1.0 and later features (v0.1 is too close now to get you started on contributions)
- advises on features (questionning the planned features for v1.0 and later)
- code examples.
- promotion on twitter, conferences and so on.

**Comment by bradstewart** [comment]
Thanks for the update!

I do intend to use this in my day job, in a setup pretty similar to what you described: a handful of shared libraries, and a handful of deployable applications which depend on those libraries. Since it's all internal code, I'd rather force all applications to use the "latest" version of those libs rather than publishing everything to npm and dealing with "real" versions.

Repo looks something like this:
```
apps
  app1
    src
    package.json
    Dockerfile
  app2
    src
    package.json
    Dockerfile
  app3
    src
    package.json
    Dockerfile
packages
  shared-lib-1
    src
    package.json
  shared-lib-2
    src
    package.json

tsconfig.json
package.json <- Holds common dev/test deps.
README.md
...etc
```

We're not using yarn at the moment, but that's not a deal-breaker for me (in fact I've been considering the move anyway, npm seems to regenerate `package-lock.json` way too often and yarn's workspaces look great).

Right now, all of those applications are Docker containers destined for embedded Linux devices (e.g. Raspberry Pi); but if things go well, I'm planning to move all of our TypeScript/JavaScript code (UI, AWS Lambda functions, etc) into this repo as well.

As for helping out, that sounds like a reasonable plan to me. I'll start poking around the codebase a bit more this week. Let me know if there's something specific I should look at.

**Comment by bradstewart** [comment]
Reading through my notes on the requirements of the tool I was about to build... might be useful to have here when it comes to feature planning (it's very similar to what you have in the README already).

- All code in one repo (organized as above).
- On `git push`, a CI server does the following:
  - Detect changed packages, and for each changed package:
    - If shared package, test and build all dependent applications. 
    - If application, test and build using latest version of all shared libs. 
  - Store application build artifacts containing:
    - Increment application version and add git tags.
    - Include a build number (could simply be git commit hash).
    - Bundled and transpiled internal/local sources.
    - Combined `package.json` file (basically flattening the relevant packages from the monorepo into a single package).
      -  Prefer not to copy shared packages as is and have to run e.g. `lerna bootstrap` to symlink everything together. Though that strategy does work for Docker applications.
    - Dockerfile, etc.
  - Trigger an application-specific release/deployment process.

I obviously don't expect monopack to deal with any of the CI pipeline (executing tests, artifact storage, deployment, etc)--just including that for context. 

I'd use monopack to:
- Detect changed packages.
- Generate the bundled/transpiled application .js files and package.json.

**Comment by flegall** [comment]
Detecting changed packages is another project I have in mind ! :)

The difficulty with performing CI on monorepos is that very often you end up building/deploying all your code on every commit and this can be quite annoying. Especially when you just modify the internal markdown documentation that is not published anywhere...

Google, Facebook and twitter use build tools to solve that problem(bazel, buck, pants).
I haven't tried those, they look a bit intimidating, but I think that a simple solution based on [jest-haste-map](https://github.com/facebook/jest/tree/master/packages/jest-haste-map) could do the job, at least for the js (+ static assets imported/required with webpack : images, css modules)

When doing cross-languages dependencies (for example: generating protobufs in C / java / js), a multi language build tool would be required.

**Comment by bradstewart** [comment]
I've been using Lerna (which works well on the development side, but not production deployments) and `lerna updated` is granular enough for my change detection needs thus far. Using `git diff` against the last successful release (however you define that) should work as well, and would let you ignore documentation changes.

I'm limiting the repo to JS code (for now, may pull in Ruby and Android stuff later). I've poked around at bazel and pants, never had a real need to dig in further. Good ol Makefiles would probably work as well, although I'm not sure how file timestamps are handled by CI servers...

**Comment by flegall** [comment]
I'm closing this, as I've set up milestones and filled issues for it !

**Comment by flegall** [comment]
I've released v0.1.0 !
